File: protocol-generator/typescript/src/codec/hash-formats.ts
import { concatBytes } from "./bytes.js";
import { decodeLeb128, encodeLeb128 } from "./leb128.js";
import { hashFormatToFunction } from "../crypto/index.js";
import { type CryptoProvider, defaultProvider } from "../crypto/index.js";
import { EntityCodecError } from "../errors.js";

/**
 * The `content_hash_format` registry (V7 §1.2). A content hash is a flat byte
 * string: a LEB128 varint *format code* followed by the digest the code names.
 * The format code is intrinsic to the hash ("interpretation", not "routing" —
 * v7.68 §1.2 reframe): byte-equality is over the full hash, format code included
 * (§5.3).
 *
 * This is the hash half of the crypto-agility seam (RESYNC v7.56→v7.70 §3).
 * Adding a hash family is a registry entry, not a rewrite. The conformance floor
 * (§9.1) is SHA-256 (`0x00`); SHA-384 (`0x01`) is validated, not required — so a
 * default Ed25519+SHA-256 peer never leaves the `0x00` path and stays
 * byte-identical to the S2 codec. Dispatch itself lives in `crypto/index`
 * ({@link hashFormatToFunction}); this module adds the protocol framing.
 */

/** ECFv1 SHA-256 — production / §9.1 floor. */
export const SHA256_FORMAT = 0x00n;

/** ECFv1 SHA-384 — validated (v7.67), not required. */
export const SHA384_FORMAT = 0x01n;

/** Reserved per §1.2 — never a valid format code. */
export const RESERVED_FORMAT = 0xffn;

/**
 * The hash-format names this peer accepts, for the §4.5 `hello.hash_formats`
 * advertisement. SHA-256 (the §9.1 floor) leads; SHA-384 is the validated
 * agility family.
 */
export const SUPPORTED_HASH_FORMAT_NAMES: readonly string[] = ["ecfv1-sha256", "ecfv1-sha384"];

/** True if `code` names a hash family this peer can interpret. */
export function isSupportedHashFormat(code: bigint): boolean {
  return code === SHA256_FORMAT || code === SHA384_FORMAT;
}

/** Negotiation name for a supported format code (inverse of the advertisement). */
export function hashFormatName(code: bigint): string {
  switch (code) {
    case SHA256_FORMAT:
      return "ecfv1-sha256";
    case SHA384_FORMAT:
      return "ecfv1-sha384";
    default:
      throw new EntityCodecError(`unsupported_content_hash_format: 0x${code.toString(16)}`);
  }
}

/**
 * Build a wire content hash: `LEB128(code) ‖ digest(code, ecfBytes)`. For the
 * default SHA-256 format this is `0x00 ‖ SHA256(ecfBytes)` — 33 bytes,
 * byte-identical to the S2 codec.
 */
export function contentHashForFormat(
  code: bigint,
  ecfBytes: Uint8Array,
  provider: CryptoProvider = defaultProvider,
): Uint8Array {
  const fn = hashFormatToFunction(code, provider);
  if (fn === undefined) {
    throw new EntityCodecError(`unsupported_content_hash_format: 0x${code.toString(16)}`);
  }
  return concatBytes(encodeLeb128(code), fn.digest(ecfBytes));
}

/**
 * Read the leading format-code varint of a wire content hash. Surfaces the
 * multi-byte LEB128 path (the `VARINT-MULTIBYTE-1` agility probe) and the
 * unsupported/reserved rejections (`unsupported_content_hash_format`).
 */
export function readFormatCode(contentHash: Uint8Array): bigint {
  return decodeLeb128(contentHash, 0).value;
}

File: protocol-generator/csharp/src/EntityCore.Protocol/Codec/HashFormats.cs
using System.Security.Cryptography;

namespace EntityCore.Protocol.Codec;

/// <summary>
/// The <c>content_hash_format</c> registry (V7 §1.2). A content hash is a flat byte
/// string: a LEB128 varint <em>format code</em> followed by the digest the code
/// names. The format code is <strong>intrinsic to the hash</strong> ("interpretation",
/// not "routing" — v7.68 §1.2 reframe): byte-equality over the full hash, format code
/// included (§5.3).
///
/// <para>
/// This is the hash half of the crypto-agility seam (RESYNC v7.56→v7.70 §3). Adding a
/// second hash family is a registry entry, not a rewrite. The conformance floor (§9.1)
/// is SHA-256 (<c>0x00</c>); SHA-384 (<c>0x01</c>) is validated, not required —
/// so a default Ed25519+SHA-256 peer never leaves the <c>0x00</c> path and stays
/// byte-identical to the S2 codec.
/// </para>
/// </summary>
internal static class HashFormats
{
    /// <summary>ECFv1 SHA-256 — production / §9.1 floor.</summary>
    public const ulong Sha256 = 0x00;

    /// <summary>ECFv1 SHA-384 — validated (v7.67), not required.</summary>
    public const ulong Sha384 = 0x01;

    /// <summary>Reserved per §1.2 — never a valid format code.</summary>
    public const ulong Reserved = 0xFF;

    /// <summary>True if <paramref name="code"/> names a hash family this peer can interpret.</summary>
    public static bool IsSupported(ulong code) => code is Sha256 or Sha384;

    /// <summary>
    /// The hash-format names this peer accepts, for the §4.5 <c>hello.hash_formats</c>
    /// advertisement. SHA-256 (the §9.1 floor) leads; SHA-384 is the validated agility
    /// family.
    /// </summary>
    public static readonly IReadOnlyList<string> SupportedNames = new[] { "ecfv1-sha256", "ecfv1-sha384" };

    /// <summary>Negotiation name for a supported format code (inverse of the advertisement).</summary>
    public static string Name(ulong code) => code switch
    {
        Sha256 => "ecfv1-sha256",
        Sha384 => "ecfv1-sha384",
        _ => throw Unsupported(code),
    };

    /// <summary>Raw digest length in bytes for a supported format code.</summary>
    public static int DigestLength(ulong code) => code switch
    {
        Sha256 => 32,
        Sha384 => 48,
        _ => throw Unsupported(code),
    };

    /// <summary>Compute the raw digest of <paramref name="data"/> under format <paramref name="code"/>.</summary>
    public static byte[] Digest(ulong code, ReadOnlySpan<byte> data) => code switch
    {
        Sha256 => SHA256.HashData(data),
        Sha384 => SHA384.HashData(data),
        _ => throw Unsupported(code),
    };

    /// <summary>
    /// Build a wire content hash: <c>LEB128(code) || Digest(code, ecfBytes)</c>. For the
    /// default SHA-256 format this is <c>0x00 || SHA256(ecfBytes)</c> — 33 bytes,
    /// byte-identical to the S2 codec.
    /// </summary>
    public static byte[] ContentHash(ulong code, ReadOnlySpan<byte> ecfBytes)
    {
        byte[] prefix = Leb128.Encode(code);
        byte[] digest = Digest(code, ecfBytes);
        var hash = new byte[prefix.Length + digest.Length];
        prefix.CopyTo(hash, 0);
        digest.CopyTo(hash.AsSpan(prefix.Length));
        return hash;
    }

    /// <summary>
    /// Read the leading format-code varint of a wire content hash. Surfaces the
    /// multi-byte LEB128 path (the <c>VARINT-MULTIBYTE-1</c> agility probe) and the
    /// unsupported/reserved rejections (<c>unsupported_content_hash_format</c>).
    /// </summary>
    public static ulong ReadFormatCode(ReadOnlySpan<byte> contentHash)
    {
        int offset = 0;
        return Leb128.Decode(contentHash, ref offset);
    }

    private static EntityCodecException Unsupported(ulong code) =>
        new($"unsupported_content_hash_format: 0x{code:x}");
}
